evap/results/tools.py

```diff
@@ -42,55 +45,75 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
         assert question.is_rating_question
         self.question = discard_cached_related_objects(copy(question))
         self.additional_text_result = additional_text_result
-
+        self.counts = None
+        self.zipped_choices = None
+        counts = OrderedDict(
+            (value, [0, name, color, value]) for (name, color, value) in self.choices.as_name_color_value_tuples()
+        )
+        counts.pop(NO_ANSWER)
+        self.colors = tuple((color for _, _, color, _ in counts.values()))
         if answer_counters is not None:
-            counts = OrderedDict((value, 0) for value in self.choices.values if value != NO_ANSWER)
             for answer_counter in answer_counters:
-                counts[answer_counter.answer] = answer_counter.count
-            self.counts = tuple(counts.values())
-        else:
-            self.counts = None
+                counts[answer_counter.answer][0] = answer_counter.count
+            self.zipped_choices = list(counts.values())
+            self.counts = tuple((count for count, _, _, _ in counts.values()))
 
     @property
-    def choices(self):
+    def choices(self) -> Choices:
         return CHOICES[self.question.type]
 
     @property
     def count_sum(self) -> int | None:
-        if not self.is_published:
+        if not is_published(self):
             return None
         return sum(self.counts)
 
     @property
-    def minus_balance_count(self) -> float:
+    def minus_balance_count(self) -> float | None:
         assert self.question.is_bipolar_likert_question
+        if not is_published(self):
+            return None
         portion_left = sum(self.counts[:3]) + self.counts[3] / 2
         return (self.count_sum - portion_left) / 2
 
     @property
     def approval_count(self) -> int | None:
         assert self.question.is_yes_no_question
-        if not self.is_published:
+        if not is_published(self):
             return None
         return self.counts[0] if self.question.is_positive_yes_no_question else self.counts[1]
 
     @property
     def average(self) -> float | None:
-        if not self.has_answers:
+        if not has_answers(self):
             return None
         return sum(grade * count for count, grade in zip(self.counts, self.choices.grades)) / self.count_sum
 
     @property
     def has_answers(self) -> bool:
-        return self.is_published and any(count != 0 for count in self.counts)
+        return has_answers(self)
 
-    @property
-    def is_published(self) -> bool:
-        return self.counts is not None
+
+class _CountedRatingResult(RatingResult):
+    counts: tuple[int, int, int, int]
+    zipped_choices: list[list[Real | StrOrPromise | str | Real]]
+
+    count_sum: int
+    average: float
+    approval_count: int
+    minus_balance_count: float
+
+
+def is_published(result: RatingResult) -> TypeGuard[_CountedRatingResult]:
+    return result.counts is not None
```

Comment:
I don't know, this all feels a bit brittle, we only have this situation because we are passing `None` for `answer_counters` if we don't want to publish them. We should probably refactor this (at some point) to have actual typing describe the state of the system and not work around it with `TypeGuard`. Why is this needed here again?

Reply:
You are right, the typeguards are not the the actual target of this PR and are not required here